gorillaz.go

```diff
@@ -30,13 +32,66 @@ type Gaz struct {
 	ViperRemoteConfig bool
 	Env               string
 	// use int32 because sync.atomic package doesn't support boolean out of the box
-	isReady           *int32
-	isLive            *int32
-	streamRegistry    *streamRegistry
-	grpcListener      net.Listener
-	grpcServerOptions []grpc.ServerOption
-	configPath        string
-	serviceAddress    string // optional address of the service that will be used for service discovery
+	isReady               *int32
+	isLive                *int32
+	streamRegistry        *streamRegistry
+	grpcListener          net.Listener
+	grpcServerOptions     []grpc.ServerOption
+	configPath            string
+	serviceAddress        string // optional address of the service that will be used for service discovery
+	streamConsumers       *streamConsumerRegistry
+	streamEndpointOptions []StreamEndpointConfigOpt
+}
+
+type streamConsumerRegistry struct {
+	sync.Mutex
+	g                 *Gaz
+	endpointsByName   map[string]*StreamEndpoint
+	endpointConsumers map[*StreamEndpoint]map[*registeredConsumer]struct{}
+}
+
+func (g *Gaz) createConsumer(endpoints []string, streamName string, opts ...ConsumerConfigOpt) (StreamConsumer, error) {
+	r := g.streamConsumers
+	r.Lock()
+	defer r.Unlock()
+	target := strings.Join(endpoints, ",")
+	e, ok := r.endpointsByName[target]
+	if !ok {
+		var err error
+		Log.Debug("Creating stream endpoint", zap.String("target", target))
+		e, err = r.g.NewStreamEndpoint(endpoints, g.streamEndpointOptions...)
+		if err != nil {
+			return nil, err
```

Comment:
should we log the error, or decorate it, so it's easier to find where things went wrong?

Reply:
I have wrapped it with pkg/errors.

Btw what is the latest canonical way to wrap errors in go, should we use pkg/errors or something else?